virttest/qemu_storage.py

```diff
@@ -1550,10 +1550,49 @@ def measure(self, target_fmt, size=None, output="human"):
         cmd_list = [self.image_cmd, "measure", ("--output=%s" % output),
                     ("-O %s" % target_fmt)]
         if size:
+            if self._secret_objects:
```

Comment:
1. _secret_objects holds all secrets string list for images defined in image_chain, i.e. this object is a collection for image itself and its backing images; 
For measure operation, refer to the manul, we might not consider its backing images, we only have to consider the image itself, i.e. we don't need to add '--object secret, xxx' for its backings,
so here we might only use _secret_objects[-1] for defining secret object for the image itself.

2. '-o xxx' is used for describing the target image options, so here -o key-secret=xxx is only used when the target image is luks(we need to calculate file size required for a luks image), so what should we do when the image itself is in luks format?
when we measure a luks image, of course, we must pass key-secret to qemu-img to let qemu-img access the image, but refer to 3, -o key-secret is used for target only, so here, we have to represent the luks image in json, e.g. 
   qemu-img measure -O xxx 'json:{xxxx}'

    if self.encryption_config.key_secret:
             image_filename = "'%s'" % get_image_json(
                 self.tag, self.params, self.root_dir)

Reply:
3. when size is given, we might not need self._secret_objects[-1], which were used for source image only, so we could remove the newly added lines of code

Reply:
> 1. _secret_objects holds all secrets string list for images defined in image_chain, i.e. this object is a collection for image itself and its backing images;
>    For measure operation, refer to the manul, we might not consider its backing images, we only have to consider the image itself, i.e. we don't need to add '--object secret, xxx' for its backings,
>    so here we might only use _secret_objects[-1] for defining secret object for the image itself.
> 2. '-o xxx' is used for describing the target image options, so here -o key-secret=xxx is only used when the target image is luks(we need to calculate file size required for a luks image), so what should we do when the image itself is in luks format?
>    when we measure a luks image, of course, we must pass key-secret to qemu-img to let qemu-img access the image, but refer to 3, -o key-secret is used for target only, so here, we have to represent the luks image in json, e.g.
>    qemu-img measure -O xxx 'json:{xxxx}'
>    if self.encryption_config.key_secret:
>    image_filename = "'%s'" % get_image_json(
>    self.tag, self.params, self.root_dir)

1. agree with you
2. If the size N is given then act as if creating a new empty image file using qemu-img create.
    If FILENAME is given then act as if converting  an  existing image file using qemu-img convert.
The details like:
(1) new image
for raw/qcow2: e.g. # qemu-img measure --output=json -O qcow2 --size 4G
for luks: e.g. # qemu-img measure --output=json -O luks --object secret,id=sec0,data=base -o key-secret=sec0 --size 20G

(2) convert an existing image
raw/qcow2 to luks:
e.g. # qemu-img measure --output=json -O luks --object secret,id=test_encrypt0,data=redhat -o key-secret=test_encrypt0 -f qcow2 original.qcow2
luks to raw/qcow2:
e.g. # qemu-img measure -O qcow2  --object secret,id=sec0,data=base --image-opts driver=luks,file.filename=original.img,key-secret=sec0
luks to luks:
e.g. # qemu-img measure --output=json -O luks --object secret,id=sec1,data=base --object secret,id=sec0,data=base -o key-secret=sec1 --image-opts driver=luks,file.filename=original.img,key-secret=sec0

Reply:
I see, I recommend that we use json to represent the source image:

# qemu-img measure -O raw --object secret,id=x,data=redhat 'json:{"driver": "luks", "file":{"driver":"file", "filename":"/home/kvm_autotest_root/images/system_image.luks"}, "key-secret":"x"}'
required size: 21474836480
fully allocated size: 21474836480

As you can see from qemu_storage, usually we use json instead of --image-opts to represent an image

Reply:
Yes, I also use json format in my codes, sorry for pasting --image-opts cmd lines to the comment.